lib/active_storage_validations/content_type_validator.rb

```diff
@@ -2,27 +2,33 @@
 
 module ActiveStorageValidations
   class ContentTypeValidator < ActiveModel::EachValidator # :nodoc:
+    include OptionProcUnfolding
+
+    AVAILABLE_CHECKS = %i[with in].freeze
+    
     def validate_each(record, attribute, _value)
-      return true if !record.send(attribute).attached? || types.empty?
+      return true unless record.send(attribute).attached?
 
+      options = unfold_procs(record, self.options, AVAILABLE_CHECKS)
+      types = authorized_types(options)
+      return true if types.empty?
+      
       files = Array.wrap(record.send(attribute))
 
-      errors_options = { authorized_types: types_to_human_format }
+      errors_options = { authorized_types: types_to_human_format(types) }
       errors_options[:message] = options[:message] if options[:message].present?
 
       files.each do |file|
-        next if is_valid?(file)
+        next if is_valid?(file, types)
 
         errors_options[:content_type] = content_type(file)
         record.errors.add(attribute, :content_type_invalid, **errors_options)
         break
       end
     end
 
-    def types
-      return @types if defined? @types
-
-      @types = (Array.wrap(options[:with]) + Array.wrap(options[:in])).compact.map do |type|
+    def authorized_types(options)
+      (Array.wrap(options[:with]) + Array.wrap(options[:in])).compact.map do |type|
```

Comment:
This is an instance method but is actually used as a helper. I would really try to make clear that these authorized types are dependent on the `record` (which they are implicitly since the options depend on the record). I would therefore rewrite this method to something like this:

```ruby
def authorized_types(record)
  with = resolve_value(record, options[:with])
  in = resolve_value(record, options[:in])
  
  (Array.wrap(with) + Array.wrap(in)).compact.map do |type|

```

This makes it clear that the authorized types depend on the `record` and you don't change the original `options`.

Reply:
Actually, the authorized types depend on the options (=flat_options now) which have to be passed in. We're using this same pattern in every other validator:
```ruby
flat_options = unfold_procs(record, self.options, AVAILABLE_CHECKS)
```
The `AVAILABLE_CHECKS` constant specifies which keys from the options should be resolved (which, in this case, are `%i[with in]`).

Reply:
I don't agree, the authorized types in the end depends on the record (and the options instance variable but that doesn't have to be passed). 

You can generate these flat options if you have the options and the record.

Reply:
I see, I just moved the proc unfolding call into the authorized_types and pass the record inside that instead of the unfolded options.

Reply:
I find this a lot clearer. Hopefully, you don't mind me pushing on this but I find it important that the lowest level of dependencies (arguments) for each function are used. This way it is harder to screw up, for example by passing the `flat_options` that were generated with a different record.